tests/PySys/local_mqtt/local_mqtt_pub_sub_retain/run.py

```diff
@@ -0,0 +1,65 @@
+from pysys.basetest import BaseTest
+
+import time
+
+"""
+Validate local publishing with retain flag and subscribing:
+
+Given a configured system
+When we publish two messages with retain flag on to test/alarms/temp_sensor
+When we start tedge sub in the background that subscribes to test/alarms/temp_sensor topic
+When we publish an empty message with retain flag on to test/alarms/temp_sensor
+Then we kill the subscriber
+Then we find the retained message in the output of the subscriber log
+"""
+
+
+class MqttPublishWithRetain(BaseTest):
+    def setup(self):
+        self.tedge = "/usr/bin/tedge"
+        self.sudo = "/usr/bin/sudo"
+      
+    def execute(self):
+        
+        pub_to_set_alarm = self.startProcess(
+            command=self.sudo,
+            arguments=[self.tedge, "mqtt", "pub", "--retain", "--qos", "1", "test/alarms/temp_sensor", "alarm msg 1"],
+            stdouterr="pub_to_set_alarm",
+        )
+
+        pub_to_set_alarm = self.startProcess(
+            command=self.sudo,
+            arguments=[self.tedge, "mqtt", "pub", "--retain", "--qos", "1", "test/alarms/temp_sensor", "alarm msg 2"],
+            stdouterr="pub_to_set_alarm",
+        )
+       
+        # wait some time before starting subscribers
+        time.sleep(1)
+
+        temp_sensor_sub = self.startProcess(
+            command=self.sudo,
+            arguments=[self.tedge, "mqtt", "sub", "test/alarms/temp_sensor"],
+            stdouterr="temp_sensor_sub",
+            background=True,
+        )
+
+        pub_to_clear_alarm = self.startProcess(
+            command=self.sudo,
+            arguments=[self.tedge, "mqtt", "pub", "--retain", "--qos", "1", "test/alarms/temp_sensor", ""],
+            stdouterr="pub_to_clear_alarm",
+        )
+
+        # wait for a while before killing the subscribers
+        time.sleep(1)
+
+        # Kill the subscriber process explicitly with sudo as PySys does
+        # not have the rights to do it
+        kill = self.startProcess(
+            command=self.sudo,
+            arguments=["killall", "tedge"],
+            stdouterr="kill_out",
+        )
+     
+    def validate(self):
+        self.assertGrep("temp_sensor_sub.out", "\[test/alarms/temp_sensor\] alarm msg 2", contains=True)
+        self.assertGrep("temp_sensor_sub.out", "\[test/alarms/temp_sensor\] ", contains=True)
```

Comment:
This validation doesn't guarantee that the "retained message was cleared". You'd have received this output even if the published message was just an empty message and *without* the `retained` one. To really test the clearing contract of retained messages, you need to have a subscriber started after the retained message is cleared and validate that it doesn't receive the cleared message at all.

Reply:
One can also improve the output of `tedge mqtt sub` to display the retain flag of the received messages.

Something along the lines:

```shell
$ tedge mqtt sub 'test/alarms/#'
[test/alarms/temp_sensor +retain] alarm msg 2
[test/alarms/temp_sensor +retain]  
```

Reply:
@albinsuresh I agree, I will change it. 
@didier-wenzek I tried subscribing the way it's mentioned above. Did not find the +retain in the output.

Reply:
I mean that you can update `tedge mqtt sub` to trace the retain flag and then use this improved version in this test.

Reply:
Using an enhanced output of `tedge mqtt sub`, tracing the retain flag, doesn't work in practice.

Indeed, even if a message is published with the retain flag, this flag will be unset when delivered to an already established subscription.

See [MQTT 3.1.1 Specification](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc385349265):

> When sending a PUBLISH Packet to a Client the Server MUST set the RETAIN flag to 1 if a message is sent as a result of a new subscription being made by a Client [MQTT-3.3.1-8]. It MUST set the RETAIN flag to 0 when a PUBLISH Packet is sent to a Client because it matches an established subscription regardless of how the flag was set in the message it received [MQTT-3.3.1-9].

Hence, there is no point to update the behavior of `tedge mqtt sub`. This will not help testing.

Reply:
yes, agreed.